### DISCUSSION TITLE
Renovate erroring on lerna monorepo with `file:` include

### DISCUSSION THREAD

**Discussion by wyardley** [renovatebot/renovate#22784] [Request Help]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

Yes

### Describe the problem

I'm not sure if this is related to our project, a transient issue with Renovate, or a lerna-specific issue, but a project which uses Lerna, and which has been working with Renovate for the most part (except for some previously reported issues never fully resolved with private deps: https://github.com/renovatebot/renovate/discussions/12078) has started failing with the following sorts of errors:

Lerna itself was last updated (to 6.6.2) back about a month ago (5/11).

```
lerna info Installing external dependencies
lerna ERR! npm install --ignore-scripts --ignore-scripts --no-audit exited 1 in '@foo/internal-service'
lerna ERR! npm install --ignore-scripts --ignore-scripts --no-audit stderr:
npm WARN tar ENOENT: no such file or directory, open '/tmp/worker/a184c1/e63f92/repos/github/someorg/reponame/services/internal-service/node_modules/.staging/@msgpackr-extract/msgpackr-extract-win32-x64-3ca2932c/node.napi.node'
npm WARN tar ENOENT: no such file or directory, open '/tmp/worker/a184c1/e63f92/repos/github/someorg/reponame/services/internal-service/node_modules/.staging/nodemon-8f0ccd23/lib/config/index.js'
npm WARN tar ENOENT: no such file or directory, open '/tmp/worker/a184c1/e63f92/repos/github/someorg/reponame/services/internal-service/node_modules/.staging/fast-xml-parser-8e0069ec/src/xmlbuilder/orderedJs2Xml.js'
npm WARN tar ENOENT: no such file or directory, open '/tmp/worker/a184c1/e63f92/repos/github/someorg/reponame/services/internal-service/node_modules/.staging/mongodb-memory-server-core-e4cbf8ae/lib/index.js'
npm WARN tar ENOENT: no such file or directory, lstat '/tmp/worker/a184c1/e63f92/repos/github/someorg/reponame/services/internal-service/node_modules/.staging/mongodb-memory-server-core-e4cbf8ae/lib/util/getos'
npm WARN tar ENOENT: no such file or directory, open '/tmp/worker/a184c1/e63f92/repos/github/someorg/reponame/services/internal-service/node_modules/.staging/mocha-cad1bab1/lib/interfaces/common.js'
npm WARN tar ENOENT: no such file or directory, open '/tmp/worker/a184c1/e63f92/repos/github/someorg/reponame/services/internal-service/node_modules/.staging/express-validator-d57739c6/src/middlewares/one-of.js'
[...]
```

Running the same bootstrap command locally seems to work as expected.

### Relevant debug logs

I can sanitize / add them as needed, but let me know which sections / details you need. The most significant stuff I see shows the same errors seen above.

I can also get you the job ID if you're able to look at our logs.

### Have you created a minimal reproduction repository?
no

**Comment by nabeelsaabna** [comment]
It will be very helpful if you are able to create a minimal reproduction repository.
also job ID can be useful.

**Reply by wyardley** [reply]
Thanks!
I think the setup is complex enough that it's going to be _really_ hard to make a minimal repro.
We even have another similar setup that's not exhibiting the problem.
`3bf049bc-9af7-464b-9cdb-4acde3d27590` is one example of a job ID

I'm not seeing it on some recent runs, so I'm wondering if there were any known / transient issues on the Renovate side that might explain this?

**Reply by wyardley** [reply]
`79b3a2f6-12b5-47da-8f2d-27b5d32bff62` is another, more recent one.

**Reply by wyardley** [reply]
I think I see the issue / cause now, though. `package.json` now has a local file include.

```
"@myorg/libs": "file:../../libs",
```

Somehow, in a Renovate PR that is _not_ operating on one of the configs that has this in it, those are getting removed from `package-lock.json` when renovate (successfully) runs, and when it's operating on dependencies that _are_ in the files that have that in dependencies, the artifact update happens.

**Reply by nabeelsaabna** [reply]
job `79b3a2f6-12b5-47da-8f2d-27b5d32bff62` ran with `node 14.21.3` and `npm 6.14.18`, are these the correct versions for you project ?

**Reply by rarkins** [reply]
My best guess would be that npm's `--package-lock-only` is not playing nicely. You might also see that that flag is dropped from the updates that work, and present in the ones which fail?

**Reply by nabeelsaabna** [reply]
a failed command:
```
docker run --rm --name=renovate_a_sidecar --label=renovate_a_child
	-v /tmp/worker/8b03da/4b036b/repos/github/ORG/REPO:/tmp/worker/8b03da/4b036b/repos/github/ORG/REPO
	-v /tmp/worker/8b03da/4b036b/cache:/tmp/worker/8b03da/4b036b/cache
	-e NPM_CONFIG_CACHE -e npm_config_store -e BUILDPACK_CACHE_DIR -e CONTAINERBASE_CACHE_DIR 
	-w /tmp/worker/8b03da/4b036b/repos/github/ORG/REPO
	ghcr.io/containerbase/sidecar 
	bash -l -c 
		install-tool node 14.21.3 && 
		install-tool npm 6.14.18 && 
		hash -d npm 2>/dev/null || true &&
		install-tool lerna 6.6.2 && 
		lerna info || echo "Ignoring lerna info failure" &&
		npm install --ignore-scripts  --no-audit && 
		lerna bootstrap --no-ci --ignore-scripts -- --ignore-scripts  --no-audit
```

**Reply by rarkins** [reply]
That one looks like there's no `--package-lock-only`

**Reply by wyardley** [reply]
So the structure of this project is a little confusing / convoluted - but at a basic level, for the purposes of this discussion, a simplified version of it looks like this:
```
├── migrations
│   ├── package-lock.json
│   └── package.json
├── package-lock.json
├── package.json
└── services
    ├── bar-service
    │   ├── package-lock.json
    │   └── package.json
    └── foo-service
        ├── package-lock.json
        └── package.json
```

If we assume that all the stuff in `services/*/package.json` has that file include, and that `migrations/*` does not, the behavior I'm seeing is:
* Dependency in `migrations` updated: Renovate "succeeds", but the `package-lock.json` of everything in `services` gets the stuff from the `file:` include dropped.
* Dependency in `services/*` updated: Renovate fails with artifact update

This does make sense with the `package-lock-only` flag being involved somehow

**Reply by wyardley** [reply]
> job 79b3a2f6-12b5-47da-8f2d-27b5d32bff62 ran with node 14.21.3 and npm 6.14.18, are these the correct versions for you project ?

Yes, as old and crusty (and EOL, I think) as those are, that's correct for most of the subprojects within this repo (for now). In general, Renovate has worked very correctly / successfully even when there are subprojects in the same repo that have newer npm / node versions and lockfiles present.

Workspaces, newer npm versions, etc., will definitely help a lot with all of this.

**Reply by rarkins** [reply]
Ultimately changes to `package-lock.json` are very likely to have been made by `npm`, not Renovate. So there's two possibilities here:
- npm has a bug (and you're unlikely to get it fixed)
- Renovate should be using different commands or parameters to control npm

Either way it's likely that you need to be the one to figure it out, due to your familiarity with the project.

**Reply by wyardley** [reply]
I think I can reproduce it locally [*edit*: but not consistently!], but I think the `package-lock` getting updated is happening from the lerna step vs. the npm one

**Reply by wyardley** [reply]
Maybe related to https://github.com/lerna/lerna/issues/2352?

**Reply by wyardley** [reply]
I will post a few things I came up with, after much hair pulling, and then close this, because I agree, not something Renovate can probably solve.

* The lib included in `libs/` had a name `@ourorg/libs` that was the same as that used in the include in other leafs' `package.json` files in the file includes. By updating that to `@ourorg/shared-libs` (or any arbitrary name), I was able to get it to work more optimally locally where running `lerna` wouldn't somehow strip out lines referencing another leaf library in the `package-lock.json` when it rebuilt it. I think it's probably a bug that lerna strips that dep when it exists as a top level name, but I'm also guessing they won't be open to fixing it with the push to move to workspaces anyway. This was causing some issues like `lerna bootstrap` not linking the package at all as well as stripping it from the package-lock.
* I also tried one of the solutions mentioned above of adding the library include in the top level `package.json` as well

With both of those fixes, I had high hope that Renovate would be happy again, but alas, it is not (`1958c2f2-f056-4a0d-b6ef-0ad9470a12c3`).
